projectile.el

```diff
@@ -2963,6 +2967,15 @@ it acts on the current project."
   (or (projectile-verify-file "Eldev" dir)
       (projectile-verify-file "Eldev-local" dir)))
 
+(defun projectile-expand-file-name-wildcard (name-pattern dir)
+  "Expand the potentially wildcard-containing NAME-PATTERN in
+DIR. If there are results expanding a wildcard, get the first
+result, otherwise expand NAME-PATTERN in DIR ignoring wildcards."
+  (let ((expanded (expand-file-name name-pattern dir)))
```

Comment:
Basically, that's the part of the code I'm concerned about from a perspective standpoint, as we are doing 2 file name expansions for everything. Not sure that's really a problem, but I expect that `file-expand-wildcards` is much slower than a regular `expand-file-name`.

Reply:
Would you prefer if we checked if `NAME-PATTERN` contained `*`, `?`, `[` or `]` before doing the second check?

Reply:
I was wondering about this. In practice will we ever have a pattern different from `*.ext`? I was thinking this way we can have the fastest check - that of the first character. I'm not sure if the `file-expand-wildcards` doesn't do such checks itself, so it might be useful to check its code.

Reply:
I looked at the implementation of `file-expand-wildcards`. It looks like the way it works is by transforming the wildcard to a regex, and then calling a C function which gets a directory listing and filters for entries matching the regex.

If the pattern doesn't have any wildcard characters, the regex construction is trivialized, but it still does a directory listing and filter, just with a simpler filter.

Reply:
Well, then I guess it'd be prudent for us to restrict the wildcards we care about only to the bare mininum - `project-name.someext` and filter which something like `string-prefix-p`.

Reply:
I thought just prefix matching was a little too restrictive and would be tricky to explain properly in the docs, so I checked for a match to `[[*?]` to decide whether to perform the wildcard expansion.
I also added a test to document this performance concern.